server.js

```diff
@@ -1987,6 +1987,95 @@ cache(function(data, match, sendBadge, request) {
+cache(function(data, match, sendBadge, request) {
+  var user = match[1];   // eg, `leafo`.
+  var module_name = match[2];   // eg, `lapis`.
+  var version = match[3];   // you can explicitly specify a version
+  var format = match[4];
+  var apiUrl = 'https://luarocks.org/manifests/' + user + '/manifest.json';
+  var badgeData = getBadgeData('luarocks', data);
+  request(apiUrl, function(err, res, buffer) {
+    if (err != null) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+    try {
+      var module_info = JSON.parse(buffer).repository[module_name];
+      var versions = Object.keys(module_info);
+      if (version && versions.indexOf(version) === -1) {
+        throw new Error('unknown version');
+      }
+    } catch (e) {
+      badgeData.text[1] = 'invalid';
+      sendBadge(format, badgeData);
+      return;
+    }
+    if (!version) {
+      if (versions.length === 1) {
+        version = versions[0];
+      } else {
+        var latestVersion = {};
+        versions.forEach(function(versionString) {
+          versionString = omitv(versionString);   // remove leading 'v'
+          var v = versionString.toLowerCase();
+          v = v.replace('-', '.');   // make revision part of version
+          var versionList = [];
+          v.split('.').forEach(function(versionPart) {
+            var parsedPart = /(\d*)([a-z]*)(\d*)/.exec(versionPart);
+            if (parsedPart[1]) {
+              versionList.push(parseInt(parsedPart[1]));
+            }
+            if (parsedPart[2]) {
+              var weight;
+              // calculate weight as a negative number
+              // 'rc' > 'pre' > 'beta' > 'alpha' > any other value
+              switch (parsedPart[2]) {
+                case 'alpha':
+                case 'beta':
+                case 'pre':
+                case 'rc':
+                  weight = (parsedPart[2].charCodeAt(0) - 128) * 100;
+                  break;
+                default:
+                  weight = -10000;
+              }
+              // add positive number, i.e. 'beta5' > 'beta2'
+              weight += parseInt(parsedPart[3]) || 0;
+              versionList.push(weight);
+            }
+          });
+          if (
+            !latestVersion.list ||   // first iteration
+            luarocksVersionCompare(versionList, latestVersion.list) > 0
+          ) {
+            latestVersion.string = versionString;
+            latestVersion.list = versionList;
+          }
+        });
+        version = latestVersion.string;
+      }
+    }
+    var color;
+    switch (version.slice(0, 3).toLowerCase()) {
+      case 'dev':
+        color = 'yellow';
+        break;
+      case 'scm':
+      case 'cvs':
+        color = 'orange';
+        break;
+      default:
+        color = 'brightgreen';
+    }
+    badgeData.text[1] = omitv(version);
+    badgeData.colorscheme = color;
```

Comment:
The way you classify versions is fairly standard. Could you point to a case where the wrong version would be selected with `latestVersion()`, but the right one would be selected with this code? If yes, is this difference specific to lua?

Otherwise, I would recommend using existing functions — more code is inevitably more bugs to fix.

```js
var version = latestVersion(versions);
var vdata = versionColor(version);
badgeData.text[1] = vdata.version;
badgeData.colorscheme = vdata.color;
```

If there is a case where the wrong function would be selected, I would recommend putting the function at the end, roughly close to the definition of `phpLatestVersion()`.